### PR TITLE
Homework 3.1 + 3.2

### PR DESCRIPTION
**Pull request by Tony-Ivanova**
I am uploading Homework 3.1 + 3.2. In it I have also included parts of 4 to make sure the logic of the orders, order history and data of orders is working correctly. 
Unfortunately I won't be able to complete all homeworks due to the fact that in a few hours I'll be going on a vacation. Nevertheless, I sincerely hope that when I get back and finish them, you won't mind having a look. 
Again, thank you for the amazing workshop and hope that we did not cause too much trouble so that one day you all might be willing to do an advanced workshop :)

### CODE REVIEW

File: app/src/main/java/com/example/etherealartefacts/ui/theme/cart/CartPage.kt [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        ) {
+            Column(
+                modifier = Modifier
+                    .padding(
+                        start = dimensionResource(id = R.dimen.row_px),
+                        end = dimensionResource(id = R.dimen.row_px),
+                    )
+                    .fillMaxSize()
+            ) {
+                if (orderedProducts.value.size == 0) {
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/cart/CartPage.kt:63]: You can use the `isEmpty` shorthand.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/cart/CartPage.kt [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+        )
+    }) {
+        Box(
+            modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center
+        ) {
+            Column(
+                modifier = Modifier
+                    .padding(
+                        start = dimensionResource(id = R.dimen.row_px),
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/cart/CartPage.kt:58]: `row_px` is a bit of an unclear name for this dimen. Is that a small, medium or large size? What does it signal?

File: app/src/main/java/com/example/etherealartefacts/ui/theme/cart/CartViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        val currentProducts = _orderedProducts.value.toMutableList()
+        currentProducts.remove(product)
+        _orderedProducts.value = currentProducts
+        updateTotalPrice()
+    }
+
+    private fun updateTotalPrice() {
+        var total = 0
+
+        for (product in _orderedProducts.value) {
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/cart/CartViewModel.kt:76]: `.forEach` is generally preferred over `for in`. Ideally, though, use `reduce` for such value accumulations as it's the most efficient.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/home/HomePage.kt [hunk 4]
```diff
@@ -1,119 +1,149 @@
...
-    val productState = remember { mutableStateOf<List<ProductWithCategory>?>(emptyList()) }
+    val sheetState = rememberModalBottomSheetState()
+    val scope = rememberCoroutineScope()
+    val filtersNumber = viewModel.filtersNumber.collectAsState()
 
-    val searchQuery = remember { mutableStateOf(TextFieldValue("")) }
-
-    LaunchedEffect(Unit) {
-        viewModel.getAllProducts()
+    LaunchedEffect(isError) {
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/home/HomePage.kt:61]: Note: You can use a `SharedFlow` of type `Unit` for this state as shown on the last lecture as to avoid unnecessary boolean value passing.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/home/HomePage.kt [hunk 5]
```diff
@@ -1,119 +1,149 @@
...
-                val isBack = false;
-                TopBar(
-                    isBack,
-                    title = stringResource(id = R.string.home),
-                    navController = navController
+    Scaffold(
+        topBar = {
+            val isBack = false
+            TopBar(
+                isBack,
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/home/HomePage.kt:81]: Why don't you just pass it as `isBack = false`? The variable up top seems unnecessary.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/home/HomeViewModel.kt [hunk 6]
```diff
@@ -21,31 +21,201 @@ class HomeViewModel @Inject constructor(
...
+        val filteredList = products.value.filter { product: Product ->
+            product.title.contains(value, ignoreCase = true)
+        }
+        _filteredProducts.value = filteredList
+    }
+
+    fun onFiltersApplied(
+        selectedStars: Int, minPrice: Int, maxPrice: Int, categories: List<Category>
+    ) {
+        if (selectedStars == lastSelectedStars && minPrice.toFloat() == lastMinPrice && maxPrice.toFloat() == lastMaxPrice && categories == lastSelectedCategories) {
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/home/HomeViewModel.kt:92]: Too long of a line. Try to stay within the IDE vertical line guide. For such large lines simply break it up into more lines or extract the conditions to several variables.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/home/HomeViewModel.kt [hunk 7]
```diff
@@ -21,31 +21,201 @@ class HomeViewModel @Inject constructor(
...
             } finally {
                 _isLoading.value = false
             }
         }
     }
+
+    init {
+        getAllProducts()
+
+        viewModelScope.launch {
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/home/HomeViewModel.kt:208]: You might want to extract this to a function also.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/product/OrderHistory.kt [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+@SuppressLint("UnusedMaterialScaffoldPaddingParameter")
+@Destination
+@Composable
+fun OrderHistory(
+    destinationsNavigator: DestinationsNavigator, cartViewModel: CartViewModel
+) {
+    val dividerDimen = dimensionResource(id = R.dimen.divider_size)
+    val orderHistory = cartViewModel.orderHistory.collectAsState()
+    val dateFormatter = SimpleDateFormat("dd.MM.yyyy", Locale.getDefault())
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/product/OrderHistory.kt:32]: Note: Might be nicer to be extracted to some Helper/Constant file/class.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/product/OrderHistory.kt [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+            Column(
+                modifier = Modifier.padding(
+                    start = dimensionResource(id = R.dimen.row_px),
+                    end = dimensionResource(id = R.dimen.row_px),
+                )
+
+            ) {
+                Column {
+
+                    for (order in orderHistory.value) {
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/product/OrderHistory.kt:57]: `.forEach` is preferable.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/product/OrderHistory.kt [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                        Row(
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/product/OrderHistory.kt:58]: Extract such repeatable views to a separate composable either in the same file for small things or into a new file in an adjacent `layouts` folder.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/product/ProductDetailsPage.kt [hunk 11]
```diff
@@ -37,66 +34,66 @@ import androidx.compose.ui.res.integerResource
...
+    val product by viewModel.productState.collectAsState()
     val isLoading by viewModel.isLoading.collectAsState()
+    val isError by viewModel.isError.collectAsState()
+    val errorMessage = stringResource(id = R.string.fetch_error)
     val context = LocalContext.current
 
-    val productState = remember { mutableStateOf<ProductWithCategory?>(null) }
-
     LaunchedEffect(Unit) {
         productId?.let { id ->
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/product/ProductDetailsPage.kt:69]: We demonstrated `SavedStateHandle` - a mechanism to supply navigation arguments directly to the view model. You can use it for this here so that this business logic doesn't have to pass through the view unnecessarily.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/product/ProductDetailsPage.kt [hunk 12]
```diff
@@ -37,66 +34,66 @@ import androidx.compose.ui.res.integerResource
...
             viewModel.getProductDetails(id)
         }
     }
 
+    LaunchedEffect(isError) {
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/product/ProductDetailsPage.kt:74]: This could be a `Unit` type `SharedFlow` also, as to avoid unnecessary boolean type checking.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/profile/ProfilePage.kt [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+import com.example.etherealartefacts.ui.theme.productDescription
+import com.example.etherealartefacts.ui.theme.topbar.TopBar
+import com.ramcosta.composedestinations.annotation.Destination
+import com.ramcosta.composedestinations.navigation.DestinationsNavigator
+
+@SuppressLint("UnusedMaterialScaffoldPaddingParameter")
+@Destination
+@Composable
+fun ProfilePage(
+    destinationsNavigator: DestinationsNavigator, cartViewModel: CartViewModel
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/profile/ProfilePage.kt:45]: Note: When you have the spare time, refactor this so that you facilitate a `DataStore` for the cart information. And if you still need to pass a view model around, either prop it up as a navigation dependency or have the store owner as a dependency and initialize it here with `hiltViewModel(viewModelStoreOwner)` as @yll090 showcased at the Q&A.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/profile/ProfileViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+@HiltViewModel
+class ProfileViewModel @Inject constructor(
+    private val repository: DefaultRepository,
+    private val jwtTokenProvider: JWTTokenProvider
+) : ViewModel() {
+
+    fun logout() {
+        viewModelScope.launch(Dispatchers.IO) {
+            jwtTokenProvider.clearToken()
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/profile/ProfileViewModel.kt:25]: Shouldn't it also navigate to the Login screen and clear the whole navigation stack?

File: app/src/main/java/com/example/etherealartefacts/ui/theme/reusable/BottomSheet.kt [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+import com.example.etherealartefacts.ui.theme.BW
+import com.example.etherealartefacts.ui.theme.Purple
+import com.example.etherealartefacts.ui.theme.home.HomeViewModel
+import com.ramcosta.composedestinations.annotation.Destination
+import kotlinx.coroutines.launch
+
+@OptIn(ExperimentalMaterialApi::class, ExperimentalMaterial3Api::class)
+@Destination
+@Composable
+fun BottomSheet(sheetState: SheetState) {
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/reusable/BottomSheet.kt:39]: Be more specific with the naming. Just a `BottomSheet` sounds generic and doesn't infer any concrete meaning.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/reusable/BottomSheet.kt [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+                ) {
+                    Text("$${updatedRange.value.start.toInt()}")
+                    Text("$${updatedRange.value.endInclusive.toInt()}")
+                }
+            }
+
+            Row(verticalAlignment = Alignment.CenterVertically) {
+                Text(text = stringResource(id = R.string.product_rating))
+
+                for (i in 1..5) {
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/reusable/BottomSheet.kt:117]: `1..5.forEach` is preferred here.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/reusable/ExpandableList.kt [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+import com.example.etherealartefacts.models.request.Category
+import com.example.etherealartefacts.ui.theme.home.HomeViewModel
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun ExpandableList(categories: List<Category>) {
+    var expanded = remember { mutableStateOf(false) }
+    val viewModel: HomeViewModel = hiltViewModel()
+    val rotateState = animateFloatAsState(
+        targetValue = if (expanded.value) 180F else 0F,
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/reusable/ExpandableList.kt:40]: Those values (180F and 0F) should be exported as constants. There is no need for `f` to be capital too.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/reusable/ExpandableList.kt [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+    )
+
+    Column(
+        modifier = Modifier.fillMaxWidth()
+    ) {
+        Text(
+            text = "Category",
+            modifier = Modifier.fillMaxWidth(0.92F),
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/reusable/ExpandableList.kt:48]: No need for `f` to be capital and can be exported as a constant (even a local one in the file if necessary).

File: app/src/main/java/com/example/etherealartefacts/ui/theme/reusable/ExpandableList.kt [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+        )
+        Card(onClick = { expanded.value = !expanded.value }) {
+            Row(
+                modifier = Modifier.fillMaxWidth(),
+                horizontalArrangement = Arrangement.SpaceBetween,
+                verticalAlignment = CenterVertically
+            ) {
+                Text(
+                    text = "All categories",
+                    modifier = Modifier.fillMaxWidth(0.92F),
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/reusable/ExpandableList.kt:59]: Same comment.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/topbar/TopBar.kt [hunk 20]
```diff
@@ -8,24 +10,36 @@ import androidx.compose.material.IconButton
...
+    destinationsNavigator: DestinationsNavigator,
+) {
+
     Column(
         modifier = Modifier
             .fillMaxWidth()
             .padding(
-                start = 15.dp,
-                end = 15.dp
+                start = dimensionResource(id = R.dimen.px_15),
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/topbar/TopBar.kt:41]: Avoid such naming as `px_15`. If you need to make it 16 some day, you will need to refactor the whole key and all files that use it. Also `px` is a bit misleading as we're using either `dp` or `sp` in Android, if I'm getting that right.

File: app/src/main/res/values/dimen_cart.xml [hunk 21]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <dimen name="cartProduct_height">56dp</dimen>
+    <dimen name="cartProduct_width">56dp</dimen>
+    <dimen name="cartProduct_curve">15dp</dimen>
+
+    <dimen name="cartProduct_title">19dp</dimen>
+    <dimen name="cartProduct_quantity">16dp</dimen>
+
+    <dimen name="emptyCart_pt">40dp</dimen>
```
Comment (gbuyukliev-accedia) [inline app/src/main/res/values/dimen_cart.xml:10]: `pt` as in points? Slightly misleading as we're using `dp`.

**Review by gbuyukliev-accedia — APPROVED** [review]
Thank you for the participation! Left a few minor comments there for you to look at when you have the time. It's no problem for us to continue reviewing all homeworks, of course.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/product/OrderHistory.kt [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+        )
+    }) {
+
+        Box(
+            modifier = Modifier.fillMaxSize(),
+
+            ) {
+            Column(
+                modifier = Modifier.padding(
+                    start = dimensionResource(id = R.dimen.row_px),
```
Comment (yll090) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/product/OrderHistory.kt:50]: Use `horizontal` instead.

File: app/src/main/java/com/example/etherealartefacts/ui/theme/profile/ProfilePage.kt [hunk 23]
```diff
@@ -0,0 +1,153 @@
...
+    Scaffold(topBar = {
+        val isBack = true
+        TopBar(
+            isBack,
+            title = stringResource(id = R.string.profile),
+            destinationsNavigator = destinationsNavigator
+        )
+    }) {
+
+        Box(
```
Comment (yll090) [inline app/src/main/java/com/example/etherealartefacts/ui/theme/profile/ProfilePage.kt:59]: Try to refactor the code a bit so it is more readable -- there are a lot of containers, indentations and is hard to follow. You can just extract some of those in this file as separate composables. This is valid for other long files, too.

**Review by yll090 — CHANGES_REQUESTED** [review]
